src/clients/grpc.rs

```diff
@@ -0,0 +1,310 @@
+/*
+ Copyright FMS Guardrails Orchestrator Authors
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+*/
+use std::future::Future;
+
+#[derive(Clone)]
+pub struct GrpcClient<C> {
+    pub service_name: String,
+    pub inner: C,
+    pub tracing_enabled: bool,
+}
+
+impl<C: Clone> GrpcClient<C> {
+    pub fn new(service_name: String, inner: C, tracing_enabled: bool) -> Self {
+        Self {
+            service_name,
+            inner,
+            tracing_enabled,
+        }
+    }
+}
+
+/// Work around for compiler bug: https://github.com/rust-lang/rust/issues/96865
+pub trait SendFuture: Future {
+    fn send(self) -> impl Future<Output = Self::Output> + Send
+    where
+        Self: Sized + Send,
+    {
+        self
+    }
+}
+
+impl<T: Future> SendFuture for T {}
+
+#[macro_export]
+macro_rules! grpc_call {
+    ($client:expr, $request:expr, $method:expr) => {{
+        use tokio::time::Instant;
+        use tracing::{debug, info};
+
+        use $crate::{tracing_utils::{current_trace_id, trace_context}};
+
+        let mut client = $client.inner.clone();
+
+        // For health clients
+        if !$client.tracing_enabled {
+            let response = $method(&mut client, $request).await?;
+            return Ok(response.into_inner())
+        }
+
+        let trace_id = current_trace_id();
+        let protocol = "gRPC";
+        let service_name = $client.service_name.clone();
+        let request_metadata = $request.metadata();
+
+        info!(?trace_id,
+            protocol,
+            service_name,
+            // method,
```

Comment:
still looking for a "clean" solution to getting the specific grpc method name in these metrics. For the most part, this information can already be found through traces (ie by looking at the instrumented function name that parent these metric info events, which would be the handler for the endpoint/method being accessed. But i imagine it would still be nice to have the specific type name of the grpc service method being called serialized here if possible